utils/cautils/acmeutils.go

```diff
@@ -494,7 +496,76 @@ func doDeviceAttestation(ctx *cli.Context, ac *ca.ACMEClient, ch *acme.Challenge
 	if err != nil {
 		return fmt.Errorf("error marshaling payload: %w", err)
 	}
-	return ac.ValidateWithPayload(ch.URL, payload)
+
+	ui.Printf("Using Device Attestation challenge to validate %q", identifier)
+	ui.Printf(" .") // Indicates passage of time.
+
+	if err := ac.ValidateWithPayload(ch.URL, payload); err != nil {
+		ui.Printf(" Error!\n\n")
+		return errors.Wrapf(err, "error validating ACME Challenge at %s", ch.URL)
+	}
+
+	var (
+		isValid = false
+		vch     *acme.Challenge
+	)
+
+	for attempts := 0; attempts < 10; attempts++ {
+		vch, err = ac.GetChallenge(ch.URL) // TODO(hs): GetChallenge should return an ACME GetChallenge client response type; not core acme.Challenge type (for safety)
+		if err != nil {
+			ui.Printf(" Error!\n\n")
+			return errors.Wrapf(err, "error retrieving ACME Challenge at %s", ch.URL)
+		}
+		// break early if challenge validation failed. Due to nature of the `device-attest-01`
+		// challenge validation, this will (currently) happen immediately, as the validation is
+		// performed synchronously by the CA when the request for `ac.ValidateWithPayload` is
+		// sent. In the (near) future we might change the way the validation is performed, so
+		// keeping the retry logic intact for now.
+		if vch.Status == "invalid" {
+			break
+		}
+		if vch.Status == "valid" {
+			isValid = true
+			break
+		}
+		ui.Printf(".")
+		time.Sleep(2 * time.Second) // NOTE: a bit shorter than the other challenge types
+	}
+	if !isValid {
+		ui.Printf(" Error!\n\n")
+		return errors.New(extractDetailedErrorMessageFromChallenge(vch))
+	}
```

Comment:
It looks like we can refactor this, and use it for http-01 challenges too. Perhaps passing the sleep duration as a parameter.

Reply:
Fixed in https://github.com/smallstep/cli/pull/837/commits/9df571805ce11deeebf860225105b717b5f4d2a0. The HTTP challenge didn't break early on `invalid`, but I think it makes sense to do it in that case too. The CA will invalidate challenges when the key authoriization isn't correct; that's not going to be fixed by repeating the request.